CSS module: Conditional rules (#31387)
mdn/content: files/en-us/web/css/css_conditional_rules/using_feature_queries/index.md

## Syntax

CSS feature queries are part of the [CSS conditional rules](/en-US/docs/Web/CSS/CSS_conditional_rules) module, which also defines the media query [`@media`](/en-US/docs/Web/CSS/@media) at-rule. You will find feature queries behave similarly to [media queries](/en-US/docs/Web/CSS/CSS_media_queries/Using_media_queries). The difference is that with a media query, you are testing something about the environment in which the web page is running, whereas with feature queries you are testing browser support for CSS features.

A feature query consists of the `@supports` at-rule followed by the support condition or a `supports()` function and declaration parameter within an `@import` at-rule declaration:

```css
@supports <support-condition> {
  CSS rules to apply
}

@import url_to_import supports(<declaration>);
```

For example, we can include a styles sheet if the user-agent supports red` as a valid value for the CSS {{cssxref("color")}} property (hint: it is):

```css
@supports (color: red) {
  CSS rules to apply
}

@import `/css/styles.css` supports(color: red);
```

> If you want to check if a browser supports the `row-gap` property, for example, you would write the following feature query. It doesn't matter which value you use in a lot of cases: if all you want is to check that the browser supports this property, then any valid value will do.

Reviewer: @estelle looks like this one still needs addressing.

}
```

> For example, if the CSS you want to run requires that the browser supports CSS Shapes and CSS Grid, you could create a rule that checks for both of these things. The following rule will only return true if `shape-outside: circle()` and `display: grid` are both supported by the browser.

Reviewer: ```suggestion
For example, if the CSS you want to run requires that the browser supports CSS Shapes and CSS Grid, you could create a rule that tests browser support for both of these features. The following rule will only return true if `shape-outside: circle()` and `display: grid` are both supported by the browser.
```

Reviewer: Also, link "CSS Shapes" and "CSS Grid" to the relevant module landing pages?

Reviewer: Also, the example is again illogical — red should indicate lack of support, not support. I'll not point this out for future examples, but make sure you keep it in mind as you look at the ones below this one and decide what to do about them.

Reviewer: @estelle I see you've updated some of these examples, which is great, but red indicating support is still a bit odd. Wouldn't green and dashed border for support be better?

{{EmbedGHLiveSample("css-examples/feature-queries/and.html", '100%', 600)}}

## Testing for at least one of multiple features

> You can also use `or`, if one property out of a selection could match to enable the CSS you want to use:

Reviewer: ```suggestion
You can also use `or` to apply CSS only if one or more declarations are supported:
```

Reviewer: Also, maybe update the below example to test for a `-webkit-` prefixed property of some kind? Testing for `-ms-grid` is completely pointless in the modern browser landscape.

Author: see https://github.com/mdn/css-examples/pull/166

Reviewer: @estelle Nice one - this example is now much more logical. However, the example has a bug in it and needs an update. As per the first note on https://developer.mozilla.org/en-US/docs/Web/CSS/font-smooth, `-webkit-font-smoothing` takes different values to the non-prefixed `font-smooth` property.

The example works if you swap out `-webkit-font-smoothing: always` for `-webkit-font-smoothing: antialiased`.

Author: fixed in https://github.com/mdn/css-examples/pull/167

Reviewer: Perfect!

## Limitations of feature queries

> The `@supports` rule tests to see if the browser can parse one or more property/value pairs, and therefore if it claims to support the feature(s). If the property and value pair is understood by the browser it returns a positive response. Feature queries checks that the declaration would be considered valid by the browser, but can not be used to check if a browser supports a thing properly, and without bugs. Feature queries cannot test for _partial implementations_.

Reviewer: ```suggestion
The `@supports` rule tests to see if browsers can parse one or more property/value pairs, and therefore if they claim to support the associated feature(s). If the property/value pairs are understood by a browser it returns a positive response. Feature queries check that declarations are considered valid by a browser, but can't be used to check if it supports a feature properly without bugs or spec violations. Feature queries cannot test for _partial implementations_.
```

We have a problem however, caused by the `width` property we used on our floated items to make them display as three columns. This is now interpreted by grid as being the width of the column track, not the width of the container as it is for the float.

{{EmbedGHLiveSample("css-examples/feature-queries/step2.html", '100%', 900)}}

What we need is a way to remove the width if `display: grid` is supported. This is exactly the situation feature queries solve. We can set the `width` back to `auto` if grid is supported.

{{EmbedGHLiveSample("css-examples/feature-queries/step3.html", '100%', 900)}}

In the above scenario, it doesn't matter that IE11 doesn't support feature queries or CSS Grid — it would get the floated version anyway, which is then overwritten by browsers that do support grid.

An alternate way to write the above code is to wrap all of the grid code in a feature query as follows.

> {{EmbedGHLiveSample("css-examples/feature-queries/step4.html", '100%', 900)}}

Reviewer: I know this example is out of date, hence you getting rid of it. However, it does feel like a bit of a shame that the article no longer has a "real" example in it. Could you maybe add back in a simpler one showing usage of a feature query, or link to some examples elsewhere?

Author: added an accent color example with both regular and "not"

> Feature queries are a useful tool when progressively enhancing a site. As we've seen, they enable you to provide a good solution for all browsers, and an enhanced solution for those browsers that support newer properties and values.

Reviewer: ```suggestion
Feature queries are a useful tool for progressively enhancing a site. They enable you to provide a good solution for all browsers, and an enhanced solution for browsers that support newer properties and values.
```

> Feature queries are not limited to property-value pairs. You can also check support for [`font-tech()`](/en-US/docs/Web/CSS/@supports#font-tech), [`font-format()`](/en-US/docs/Web/CSS/@supports#font-format), and [`selector()`](/en-US/docs/Web/CSS/@supports#function_syntax).

Reviewer: I'd update the link text in these links to describe in words what is being tested for, not just list the functions used to write the necessary support queries, so for example "font technologies", "font formats", and "selectors". When I first read this I went, for example, "hrm, I wonder what the selector() function is and why I'd want to test for that?" It took me a few seconds to realize that you just meant selectors.

Author: moved up into a "## Additional feature query options" section and added a faux example.

Reviewer: Great! This works much better now.

{{EmbedGHLiveSample("css-examples/feature-queries/or.html", '100%', 600)}}

## Additional feature query options

Feature queries are not limited to property-value pairs. You can include [`font-tech()`](/en-US/docs/Web/CSS/@supports#font-tech), [`font-format()`](/en-US/docs/Web/CSS/@supports#font-format), and [`selector()`](/en-US/docs/Web/CSS/@supports#function_syntax) functions in your feature query to target CSS based on whether the user-agent supports a specified font technology, font format, or selector syntax, respectively.

For example, the `selector()` function can be used to import a stylesheet for browsers that support a vendor prefixed pseudo-element:

```css
/* a `selector()` query within a `supports()` function */
@import `/css/webkitShadowStyles.css` supports(selector(::-webkit-inner-spin-button));
```

## Examples

### Browser support test

> In this example, we check if the browser supports {{cssxref("system-color")}}, hiding support with `diplay: none` if the color type is not supported.

Reviewer: ```suggestion
In this example, we check if the browser supports the `AccentColor` {{cssxref("system-color")}}, hiding support with `display: none` if the color type is not supported.
```
